test/dispatch.jl

```diff
@@ -0,0 +1,91 @@
+@testset "type_morespecific" begin
+    @testset "subtype" begin
+        @test type_morespecific(Int, Integer)
+        @test !type_morespecific(Integer, Int)
+        @test !type_morespecific(Int, Int)
+    end
+
+    @testset "number of arguments" begin
+        @test type_morespecific(Tuple{Int,AbstractFloat}, Tuple{Integer})
+    end
+
+    @testset "ambiguous" begin
+        @test !type_morespecific(Tuple{Int,AbstractFloat}, Tuple{Integer,Float64})
+        @test !type_morespecific(Tuple{Integer,Float64}, Tuple{Int,AbstractFloat})
+    end
+
+    @testset "length" begin
+        @test type_morespecific(Tuple{Int,Int}, Tuple{Int,Vararg{Int}})
+    end
+end
+
+@testset "anonymous_signature" begin
+    plus_method = first(methods(+, (Int, Int)))
+    plus_sig = Tuple{typeof(+),T,T} where T <: Base.BitInteger
+    @test plus_method.sig == plus_sig
+
+    @test anonymous_signature(Tuple{typeof(+),Int,Int}) == Tuple{Int,Int}
+    @test anonymous_signature(plus_sig) == Tuple{T,T} where T <: Base.BitInteger
+    @test anonymous_signature(plus_method) == Tuple{T,T} where T <: Base.BitInteger
+end
```

Comment:
Can we havde a more complex test involving `where`?

Perhaps:

```
julia> methods(adjoint).ms[5]
adjoint(S::LinearAlgebra.SymTridiagonal{#s571,V} where V<:AbstractArray{#s571,1} where #s571<:Real) in LinearAlgebra at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.0/LinearAlgebra/src/tridiag.jl:137
```

Reply:
I'll add something. I was slightly reluctant to use a real-world function as the signature could change and break the test

Reply:
we could reasonably write one similr to that

Reply:
I've revised the tests to specifically test for diagonal and triangular dispatch